### PR TITLE
[Feature] Admin User Specific Spam

### PR DESCRIPTION
**Pull request by Ghalko**
## Proposal
Connect user profile to spam list so that an administrator can see the flagged/marked/ham from the user in order to evaluate the user's spam potential. ADM-62

## Repercussions
More connectivity == more power

## Order
After #4951

### CODE REVIEW

File: admin/spam/views.py [hunk 1]
```diff
@@ -1,47 +1,142 @@
...
+        self.page = request.GET.get('page', 1)
+        context['page_number'] = self.page
+        self.status = request.GET.get('status', 1)
+        context['status'] = self.status
+        context['form'] = self.get_form()
+        if context['form'].is_valid():
+            return self.form_valid(context['form'])
+        else:
+            return render(request, self.template_name, context=context)
+        # return super(SpamDetail, self).post(request, *args, **kwargs)
```
Comment (mfraezz) [inline admin/spam/views.py]: Please remove this comment

File: admin/spam/views.py [hunk 2]
```diff
@@ -1,47 +1,142 @@
...
+
+    def get_context_data(self, **kwargs):
+        self.spam_id = kwargs['spam_id']
```
Comment (mfraezz) [inline admin/spam/views.py]: Could this throw an uncaught `KeyError`?

File: admin/templates/nodes/node.html [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                    {% endfor %}
+                    </tbody>
+                    </table>
+                </td>
+            </tr>
+        </tbody>
+        </table>
+        </div>
+    </div>
+{% endblock content %}
```
Comment (mfraezz) [inline admin/templates/nodes/node.html]: Newline please

File: admin_tests/pre_reg/test_views.py [hunk 4]
```diff
@@ -1,110 +0,0 @@
...
-from website.models import MetaSchema
-from website.project.model import ensure_schemas
-
-from admin.common_auth.models import MyUser as User
-from admin.pre_reg.views import *  # noqa
-
-PREREG_GROUP = Group.objects.get(name='prereg_group')
-
-
-class PreregViewsTests(AdminTestCase):
```
Comment (mfraezz) [inline admin_tests/pre_reg/test_views.py]: Why was this file removed?

File: admin/base/forms.py [hunk 5]
```diff
@@ -0,0 +1,5 @@
+from django import forms
+
+
+class GuidForm(forms.Form):
+    guid = forms.CharField(label='Guid', max_length=5, required=True)
```
Comment (mfraezz) [inline admin/base/forms.py]: This should probably specify `min_length=5` too, unless we have any guids less than length 5 (Institutions, maybe?).

Also, it might be good to note that `max_length` will need to be updated whenever we bump the guid length up to 6.

File: admin_tests/pre_reg/test_views.py (hunk 4 above)
Comment (Ghalko) [inline admin_tests/pre_reg/test_views.py]: It didn't actually run and wasn't included in Travis for that reason. I commented it out, but it was pointed out that we could just find it in the history of commits and just to delete it for now.

File: admin/templates/spam/spam.html [hunk 6]
```diff
@@ -1,5 +1,6 @@
 {% extends "base.html" %}
```
Comment (mfraezz) [inline admin/templates/spam/spam.html]: This file might make more sense named `spam/spam_list.html`. The naming convention in other template directories is that `<model_name>.html` is the view for a `<model>` instance, whereas this is the view for a list of spam.

File: admin/templates/users/spam.html [hunk 7]
```diff
@@ -0,0 +1,85 @@
+{% extends "base.html" %}
```
Comment (mfraezz) [inline admin/templates/users/spam.html]: Would it make more sense for this file to be `/templates/spam/user.html`? We have a `/spam/comment.html` template, not `/comments/spam.html`.

File: admin/templates/util/pagination.html [hunk 8]
```diff
@@ -3,7 +3,7 @@
     <div class="pagination pagination-lg">
         <span>
             {% if items.has_previous %}
-                <a href="?page={{ items.previous_page_number }}"
+                <a href="?page={{ items.previous_page_number }}&status={{ status }}"
```
Comment (mfraezz) [inline admin/templates/util/pagination.html]: I'm not sure if this template has implications for non-spam views (maybe pre-reg), but will status always be passed in here?

File: admin/users/serializers.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
+"""
+Serialize user
+"""
+# from admin.nodes.serializers import serialize_simple_user
```
Comment (mfraezz) [inline admin/users/serializers.py]: Please remove.

File: admin_tests/pre_reg/test_views.py (hunk 4 above)
Comment (mfraezz) [inline admin_tests/pre_reg/test_views.py]: :+1:

File: admin_tests/nodes/test_views.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+            view.get_object()
+
+    def test_load_data(self):
+        node = NodeFactory()
+        guid = node._id
+        request = RequestFactory().get('/fake_path')
+        view = NodeView()
+        view = setup_view(view, request, guid=guid)
+        res = view.get_object()
+        assert_is_instance(res, dict)
```
Comment (mfraezz) [inline admin_tests/nodes/test_views.py]: Please add a few more checks here to ensure you get back the expected result, not just a `dict`. Likewise for other tests of this form.

**Comment by mfraezz** [comment]
Looks good so far, just a few minor issues that need to be addressed. Manual testing revealed no bugs :star: 

Back to you, @Ghalko

File: admin/spam/views.py (hunk 1 above)
Comment (Ghalko) [inline admin/spam/views.py]: Removed

File: admin/spam/views.py (hunk 2 above)
Comment (Ghalko) [inline admin/spam/views.py]: Made into a get with default `None`

File: admin/base/forms.py (hunk 5 above)
Comment (Ghalko) [inline admin/base/forms.py]: Done.

File: admin/templates/users/spam.html (hunk 7 above)
Comment (Ghalko) [inline admin/templates/users/spam.html]: files moved/renamed

File: admin/users/serializers.py (hunk 9 above)
Comment (Ghalko) [inline admin/users/serializers.py]: removed

File: admin_tests/nodes/test_views.py (hunk 10 above)
Comment (Ghalko) [inline admin_tests/nodes/test_views.py]: This is tested in test_serializers.py

File: admin/templates/util/pagination.html (hunk 8 above)
Comment (Ghalko) [inline admin/templates/util/pagination.html]: I think this is best left for when we look into refactoring the prereg into a CBV rather than FBV. Definitely a worthwhile consideration.

File: admin_tests/nodes/test_views.py (hunk 10 above)
Comment (mfraezz) [inline admin_tests/nodes/test_views.py]: You're right. I think I thought this was spam-related, for some reason.

On that note: why are there no spam tests?

File: admin/templates/util/pagination.html (hunk 8 above)
Comment (mfraezz) [inline admin/templates/util/pagination.html]: :+1:

**Comment by mfraezz** [comment]
Please add a few `spam` tests. If not ones for coverage of the whole module, at least for this feature.

**Comment by Ghalko** [comment]
Writing tests for spam in #5045, I would rather not write the same tests for both of these PRs.

**Comment by mfraezz** [comment]
LGTM. Putting this in RTM :crocodile:

**Comment by sloria** [comment]
@Ghalko Can you please get this up-to-date with develop one last time? Once that is done, this is good to merge.

File: admin/base/views.py [hunk 11]
```diff
@@ -74,3 +57,15 @@ def form_valid(self, form):
...
+
+class GuidView(DetailView):
+    def __init__(self):
+        self.guid = None
+        super(GuidView, self).__init__()
+
+    def get(self, request, *args, **kwargs):
+        try:
+            return super(GuidView, self).get(request, *args, **kwargs)
+        except AttributeError:
```
Comment (sloria) [inline admin/base/views.py]: Why is an `AttributeError` expected here?

File: admin/nodes/urls.py [hunk 12]
```diff
@@ -3,5 +3,8 @@
 from . import views
 
 urlpatterns = [
-    url(r'^$', views.NodeFormView.as_view(), name='node'),
+    url(r'^$', views.NodeFormView.as_view(),
+        name='search'),
+    url(r'^id-(?P<guid>[a-z0-9]+)/$', views.NodeView.as_view(),
```
Comment (sloria) [inline admin/nodes/urls.py]: Why is the `id-` prefix necessary?

File: admin/spam/urls.py [hunk 13]
```diff
@@ -3,7 +3,10 @@
 from . import views
 
 urlpatterns = [
-    url(r'^$', views.spam_list, name='spam'),
-    url(r'^(?P<spam_id>[a-z0-9]+)/$', views.spam_detail, name='detail'),
-    url(r'^(?P<spam_id>[a-z0-9]+)/email/$', views.email, name='email'),
+    url(r'^$', views.SpamList.as_view(), name='spam'),
+    url(r'^user/id-(?P<user_id>[a-z0-9]+)/$', views.UserSpamList.as_view(),
```
Comment (sloria) [inline admin/spam/urls.py]: Again, the `id-` prefix seems unnecessary.

File: admin/spam/views.py [hunk 14]
```diff
@@ -1,47 +1,141 @@
...
-        'comment': serialize_comment(Comment.load(spam_id)),
-        'page_number': request.GET.get('page', 1),
-    }
-    return render(request, 'spam/comment.html', context)
+
+class UserSpamList(SpamList):
+    template_name = 'spam/user.html'
+
+    def __init__(self):
+        self.user_id = None
```
Comment (sloria) [inline admin/spam/views.py]: Why is it necessary to set `user_id` as an attribute; you can always access it from `self.kwargs`.

Comment (sloria) [inline admin/spam/views.py]: I'm seeing this pattern in other views, too. It seems unnecessary to override `__init__` in most cases.

File: admin/nodes/urls.py (hunk 12 above)
Comment (Ghalko) [inline admin/nodes/urls.py]: It really isn't. Something that I have picked up, but it is across so many PRs at this point... I would rather change them all at once, along with a good restructuring of many urls.

File: admin/spam/views.py (hunk 14 above)
Comment (Ghalko) [inline admin/spam/views.py]: Started fixing this in the PRs that I update.

File: admin/base/views.py (hunk 11 above)
Comment (Ghalko) [inline admin/base/views.py]: This is where the class will actually return a page_not_found, either in this PR or in subsequent ones.

**Comment by Ghalko** [comment]
@sloria Responded to your concerns. Looks like this is up-to-date and all checks passed.

File: admin/nodes/urls.py (hunk 12 above)
Comment (sloria) [inline admin/nodes/urls.py]: :+1:
